service-mirror.go

```diff
@@ -0,0 +1,143 @@
+package main
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	apiError "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/selection"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/client-go/util/homedir"
+)
+
+func main() {
+	// Set up logger .
+	log := &logrus.Logger{
+		Out:   os.Stderr,
+		Level: logrus.DebugLevel,
+		Formatter: &logrus.TextFormatter{
+			TimestampFormat: "2006-01-02 15:04:05",
+			ForceColors:     true,
+			DisableColors:   false,
+			FullTimestamp:   true,
+		},
+	}
+	//Shows line number: Too long
+	// log.SetReportCaller(true)
+
+	log.Infof("Starting Global Mirror")
+
+	var kubeconfig *string
+	if home := homedir.HomeDir(); home != "" {
+		kubeconfig = flag.String("kubeconfig", filepath.Join(home, ".kube", "config"), "(optional) absolute path to the kubeconfig file")
+	}
+
+	flag.Parse()
+
+	// use the current context in kubeconfig
+	config, err := clientcmd.BuildConfigFromFlags("", *kubeconfig)
+	if err != nil {
+		log.Panicf("Probably running Inside Cluster :%s", err.Error())
+	}
+
+	// creates the clientset
+	client, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		log.Panicf("Issue in building client from config : %s", err.Error())
+	}
+
+	/* Get running services in all the names, labelled with "mirror.linkerd.io/mirrored-service: true"
+	and which service which does have label : mirror.linkerd.io/headless-mirror-svc-name. Build selector for it.
+	*/
+	mirrorSvcLabel, err := labels.NewRequirement("mirror.linkerd.io/mirrored-service", selection.Equals, []string{"true"})
+	if err != nil {
+		log.Errorf("Unable to generate error requirement : %s", err.Error())
+	}
+
+	// Create label requirements for "mirror.linkerd.io/headless-mirror-svc-name" not existing
+	mirrorSvcParentLabel, err := labels.NewRequirement("mirror.linkerd.io/headless-mirror-svc-name", selection.DoesNotExist, []string{})
+	if err != nil {
+		log.Errorf("Unable to generate error requirement : %s", err.Error())
+	}
+
+	// Create the label selector
+	svcFilter := labels.NewSelector()
+	svcFilter = svcFilter.Add(*mirrorSvcParentLabel)
+	svcFilter = svcFilter.Add(*mirrorSvcLabel)
```

Comment:
This is good but I wonder if they have a builder or something so we can do

```
svcFilter := labels.NewSelector().With(*mirrorSvcParentLabel).With(*mirrorSvcLabel)
```

I'm not sure they do but imo they should have some utils for building this stuff.

Reply:
not a blocker btw

Reply:
Yeah I will be on lookout for this, but after writing rust I was looking for this in Golang but i didnt find anything. But my guess is it smh related to generics ?